src/hotspot/cpu/riscv/relocInfo_riscv.cpp

```diff
@@ -55,7 +55,10 @@ void Relocation::pd_set_data_value(address x, bool verify_only) {
       bytes = MacroAssembler::pd_patch_instruction_size(addr(), x);
       break;
   }
-  ICache::invalidate_range(addr(), bytes);
+
+  if (!UseCtxFencei) {
+    ICache::invalidate_range(addr(), bytes);
+  }
```

Comment:
One more question: Do we need a full fence (`OrderAccess::fence()`) here with `UseCtxFencei` after the patching? Like you do in `ZBarrierSetAssembler::patch_barrier_relocation()`:
```
  if (!UseCtxFencei) {
    // A full fence is generated before icache_flush by default in invalidate_word
    ICache::invalidate_range(addr, bytes);
  } else {
    OrderAccess::fence();
  }
```

Reply:
I actually didn't look at that.
As the old case did a full fence when calling ICache::invalidate_range(addr, bytes); and with that comment,
I assumed there was a reason for pointing it, so I just kept old behavior.
I'll see If I can figure out if/why/what.

Reply:
Sorry, now I realize what you are asking.

Reply:
The purpose is to to make a store to instruction memory visible to all RISC-V harts.
Check this:
```
static int icache_flush(address addr, int lines, int magic) {
  // To make a store to instruction memory visible to all RISC-V harts,
  // the writing hart has to execute a data FENCE before requesting that
  // all remote RISC-V harts execute a FENCE.I.

  // We need to make sure stores happens before the I/D cache synchronization.
  __asm__ volatile("fence rw, rw" : : : "memory");                                                     <==============

  RiscvFlushIcache::flush((uintptr_t)addr, ((uintptr_t)lines) << ICache::log2_line_size);

  return magic;
}
```

Reply:
Yes. The comment is a bit misleading.
A fence wr, wr is just for locally, on a hart, order instructions.
A fence wr wr do not flush a store from the store buffer, it just means that the hart must act such that the store appears to have happened before the syscall. Meaning the store can not happen after the syscall.

As we are not using the syscall, instead the other threads/harts must emit the fence.i them self, either by leaving a safepoint, hitting the patch_epoch_barrier and if they are moved to another hart.

Which means the store must happen before we leave the safepoint and disarming the nmethod barrier.
These already have store fences as we have a bunch of other stores which must be ordered.
For the nmethod barrier disarming is a full fence as the new epoch must happen before the disarm:

```
(nmethod barriers armed, implies storestore)
stores to instruction stream
store new patching epoch
storestore // the store to instruction and patching epoch must happen before disarm
disarm
```

Reply:
Code would in BarrierSetNMethod::set_guard_value(...)

Reply:
I see. Seems deserve a code comment. Also, does that mean the `OrderAccess::fence()` you added in `ZBarrierSetAssembler::patch_barrier_relocation()` unnecessary?

Reply:
Yes. I added due to the comment.
I'll remove it and update that comment and add new comment about above.